### PR TITLE
Union implementation

### PR DESCRIPTION
**Pull request by nishnik**
`Union` for `Set` module.

### CODE REVIEW

File: symengine/sets.h [hunk 1]
```diff
@@ -199,5 +225,42 @@ inline RCP<const Set> interval(const RCP<const Number> &start,
         return finiteset({start});
     return emptyset();
 }
+
+// ! \return RCP<const Set>
+inline RCP<const Set> sym_union(const set_basic &in, bool solve = true)
```
Comment (isuruf) [inline symengine/sets.h]: What is the purpose of `solve` variable ?

Comment (nishnik) [inline symengine/sets.h]: `solve` variable ensures that we don't fall in a loop when we have solved the `set_union` from respective containers and they are trying to return a `Union` object.

File: symengine/tests/basic/test_sets.cpp [hunk 2]
```diff
@@ -54,18 +56,21 @@ TEST_CASE("Interval : Basic", "[basic]")
...
     r3 = interval(integer(21), integer(22));
     r4 = r1->set_intersection(r3);
     REQUIRE(eq(*r4, *emptyset()));
     r3 = interval(im5, i2, false, false); // [-5, 2]
     r4 = interval(integer(3), i20, false, false);
     REQUIRE(r3->compare(*r4) == -1);
-    CHECK_THROWS_AS(r3->set_union(r4), std::runtime_error);
-    CHECK_THROWS_AS(r4->set_union(r3), std::runtime_error);
+    REQUIRE(eq(*r3->set_union(r4), *sym_union({r3, r4})));
+    REQUIRE(eq(*r4->set_union(r3), *sym_union({r3, r4})));
```
Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: Check that `r3->set_union(r4)` is indeed a `Union` and its args are correct.

File: symengine/sets.h [hunk 3]
```diff
@@ -199,5 +225,42 @@ inline RCP<const Set> interval(const RCP<const Number> &start,
...
+    set_basic combined_FiniteSet;
+    for (auto it = in.begin(); it != in.end(); ++it) {
+        if (is_a<FiniteSet>(**it)) {
+            set_basic container;
+            const FiniteSet &other = static_cast<const FiniteSet &>(**it);
+            std::set_union(combined_FiniteSet.begin(), combined_FiniteSet.end(),
+                           other.container_.begin(), other.container_.end(),
+                           std::inserter(container, container.begin()),
+                           RCPBasicKeyLess{});
+            combined_FiniteSet = container;
```
Comment (isuruf) [inline symengine/sets.h]: `combined_FiniteSet.insert(other.container_.begin(), other.container_.end())` is better here.

File: symengine/sets.h [hunk 4]
```diff
@@ -199,5 +225,42 @@ inline RCP<const Set> interval(const RCP<const Number> &start,
...
+    } else if (input.size() == 1 && combined_FiniteSet.empty()) {
+        return rcp_dynamic_cast<const Set>(*input.begin());
+    }
+    // Now we rely on respective containers' own rules
+    // TODO: Improve it to O(log n)
+    RCP<const Set> combined_Rest = finiteset(combined_FiniteSet);
+    for (auto it = input.begin(); it != input.end(); ++it) {
+        RCP<const Set> other = rcp_dynamic_cast<const Set>(*it);
+        combined_Rest = combined_Rest->set_union(other);
+    }
```
Comment (isuruf) [inline symengine/sets.h]: Move the above for loop to a new method `interval_union`. And at the end, interval union and `combined_Finiteset` can be merged together.

**Comment by isuruf** [comment]
How does this handle `[0, 3] U [4, 5] U [3, 4]` ?

File: symengine/sets.h (hunk 4 above)
Comment (isuruf) [inline symengine/sets.h]: On second thought, `input` doesn't need to be `interval`s right? Then this is fine

**Comment by isuruf** [comment]
Also, how does `[0, 1] U [3, 4] U [2, 3]` work?

File: symengine/sets.h (hunk 4 above)
Comment (nishnik) [inline symengine/sets.h]: Yes. `input` doesn't need to be `interval`

**Comment by nishnik** [comment]
`[0, 3] U [4, 5] U [3, 4]` did work.
`[0, 1] U [3, 4] U [2, 3]` failed as the algorithm for `Union::set_union` was wrong.

**Comment by isuruf** [comment]
Can you explain how `[0, 1] U [3, 4] U [2, 3]` works step by step ?

**Comment by nishnik** [comment]
```
in `SymEngine::set_union()`:
    At line 252, 
        combined_Rest = emptyset();
        input = {interval(0, 1), interval(3, 4), interval(2, 3)}
    We enter the loop:
    First iteration:
        other = interval(0,1)
        combined_Rest = combined_Rest(= emptyset) union with other = other = interval(0, 1) //// done in Interval::set_union
    Second iteration:
        other = interval(3,4)
        combined_Rest = combined_Rest(= interval(0, 1)) union with other(= interval(3,4)) = Union(interval(3,4), interval(0, 1)) // done in Interval::set_union
    Third iteration:
        other = interval(2, 3)
        combined_Rest = combined_Rest(Union(...)) union with other = 
            In Union::set_union(const RCP<const Set> &o)
            container = {interval(3,4), interval(0,1)}
            now we try to do union between `o` and every element of container
                if it is possible to do so, we do the union and erase
                the element we have done union with and insert the union's result to container.
                and return SymEngine::sym_union(container)
                else if there is no union possible with any element, we return a new Union object, i.e.
                Union(interval(3,4), interval(0,1), interval(2, 3))
```

File: symengine/sets.cpp [hunk 5]
```diff
@@ -418,57 +378,88 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
...
-    return (*o).is_subset(rcp_from_this_cast<const Set>());
+    return SymEngine::set_union(container);
 }
 
-bool FiniteSet::is_proper_superset(const RCP<const Set> &o) const
+RCP<const Boolean> Union::contains(const RCP<const Basic> &o) const
 {
-    return (*o).is_subset(rcp_from_this_cast<const Set>()) and (not __eq__(*o));
+    for (auto &a : container_) {
+        RCP<const Set> me = rcp_dynamic_cast<const Set>(a);
```
Comment (isuruf) [inline symengine/sets.cpp]: You can use `rcp_static_cast`

File: symengine/sets.cpp [hunk 6]
```diff
@@ -418,57 +378,88 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
...
-    return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    if (is_a<UniversalSet>(*o) || is_a<EmptySet>(*o) || is_a<Union>(*o)) {
+        return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    }
+    throw std::runtime_error("Not implemented");
 }
 
-bool FiniteSet::is_subset(const RCP<const Set> &o) const
+Union::Union(set_basic in) : container_(in)
 {
```
Comment (isuruf) [inline symengine/sets.cpp]: Can you add a is_canonical check to see whether `in` contains `Set` objects?

File: symengine/sets.cpp [hunk 7]
```diff
@@ -418,57 +378,88 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
...
     }
-    return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    if (is_a<UniversalSet>(*o) || is_a<EmptySet>(*o) || is_a<Union>(*o)) {
+        return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    }
+    throw std::runtime_error("Not implemented");
 }
 
-bool FiniteSet::is_subset(const RCP<const Set> &o) const
+Union::Union(set_basic in) : container_(in)
```
Comment (isuruf) [inline symengine/sets.cpp]: It's better to use a `std::set<RCP<const Set>>` because you don't need other types here.

File: symengine/sets.h [hunk 8]
```diff
@@ -16,11 +18,23 @@ class Set : public Basic
...
-    virtual bool is_superset(const RCP<const Set> &o) const = 0;
-    virtual bool is_proper_superset(const RCP<const Set> &o) const = 0;
+    virtual RCP<const Boolean> contains(const RCP<const Basic> &a) const = 0;
+    bool is_subset(const RCP<const Set> &o) const
+    {
+        return eq(*this->set_intersection(o), *this);
+    }
+    bool is_proper_subset(const RCP<const Set> &o) const
+    {
+        return eq(*this->set_intersection(o), *this) and not eq(*this, *o);
```
Comment (isuruf) [inline symengine/sets.h]: Swap the conditions as the second one is cheaper.

File: symengine/logic.h [hunk 9]
```diff
@@ -7,6 +7,13 @@
 #define SYMENGINE_LOGIC_H
 
 #include <symengine/basic.h>
+namespace SymEngine
+{
+class Set;
+class BooleanAtom;
+class Boolean;
+inline RCP<const BooleanAtom> boolean(bool b);
+}
```
Comment (isuruf) [inline symengine/logic.h]: Move this to `sets.h`

File: symengine/logic.h [hunk 10]
```diff
@@ -7,6 +7,13 @@
 #define SYMENGINE_LOGIC_H
 
 #include <symengine/basic.h>
+namespace SymEngine
+{
+class Set;
+class BooleanAtom;
+class Boolean;
+inline RCP<const BooleanAtom> boolean(bool b);
```
Comment (isuruf) [inline symengine/logic.h]: Remove inline keyword

**Comment by isuruf** [comment]
Can you remove `symengine/tests/logic/test_logic` from history? Don't use `git add .` when you use git. Also, it's better to use a build directory.

File: symengine/printer.cpp [hunk 11]
```diff
@@ -181,6 +181,16 @@ void StrPrinter::bvisit(const EmptySet &x)
...
 }
 
+void StrPrinter::bvisit(const Union &x)
+{
+    std::ostringstream s;
+    s << **x.container_.begin();
+    for (auto it = ++(x.container_.begin()); it != x.container_.end(); ++it) {
+        s << " U " << **it;
+    }
+    str_ = s.str();
```
Comment (isuruf) [inline symengine/printer.cpp]: Can you add a test for this?

File: symengine/printer.cpp [hunk 12]
```diff
@@ -181,6 +181,16 @@ void StrPrinter::bvisit(const EmptySet &x)
     str_ = "EmptySet";
 }
 
+void StrPrinter::bvisit(const Union &x)
+{
+    std::ostringstream s;
+    s << **x.container_.begin();
+    for (auto it = ++(x.container_.begin()); it != x.container_.end(); ++it) {
+        s << " U " << **it;
```
Comment (isuruf) [inline symengine/printer.cpp]: Don't use `<< **it;` as it will prevent extending the priitnter. Use ` << apply(*it);`

File: symengine/sets.cpp [hunk 13]
```diff
@@ -150,7 +149,11 @@ RCP<const Set> Interval::set_intersection(const RCP<const Set> &o) const
             return emptyset();
         }
     }
-    return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    if (is_a<UniversalSet>(*o) or is_a<EmptySet>(*o) or is_a<FiniteSet>(*o)
+        || is_a<Union>(*o)) {
```
Comment (isuruf) [inline symengine/sets.cpp]: `||` -> `or`

File: symengine/sets.cpp [hunk 14]
```diff
@@ -184,35 +188,11 @@ RCP<const Set> Interval::set_union(const RCP<const Set> &o) const
...
     }
-    return (*o).set_union(rcp_from_this_cast<const Set>());
-}
-
-bool Interval::is_subset(const RCP<const Set> &o) const
-{
-    if (is_a<Interval>(*o)) {
-        return this->__eq__(*this->set_intersection(o));
+    if (is_a<UniversalSet>(*o) or is_a<EmptySet>(*o) or is_a<FiniteSet>(*o)
+        || is_a<Union>(*o)) {
```
Comment (isuruf) [inline symengine/sets.cpp]: `||` -> `or`

**Comment by isuruf** [comment]
Looks good to me. One more thing is to remove `symengine/tests/logic/test_logic` from history. Even if it is deleted, it is still there in the git history increasing the size of the git repo. See, https://help.github.com/articles/remove-sensitive-data/#using-filter-branch

**Comment by nishnik** [comment]
@isuruf Done!

**Pull request merged**